index.bs

```diff
@@ -11302,6 +11306,147 @@ The [=remote end steps=] given |session| and |command parameters| are:
 </div>
 
 
+## The extensions Module ## {#module-extensions}
+
+The <dfn export for=modules>extensions</dfn> module contains functionality for
+managing and interacting with extensions.
+
+### Definition ### {#module-extensions-definition}
+
+[=remote end definition=]
+
+<pre class="cddl remote-cddl">
+ExtensionsCommand = (
+  extensions.Install
+)
+</pre>
+
+[=local end definition=]
+
+<pre class="cddl local-cddl">
+ExtensionsResult = (
+  extensions.InstallResult
+)
+</pre>
+
+### Types ### {#module-extensions-types}
+
+#### The extensions.Extension Type #### {#type-extensions-Extension}
+
+<pre class="cddl local-cddl">
+extensions.ExtensionArchive = {
+   files: [+text],
+}
+</pre>
+
+<pre class="cddl remote-cddl local-cddl">
+extensions.Extension = {
+   installPath: extensions.ExtensionPath / extensions.ExtensionArchivePath / extensions.ExtensionBase64Encoded,
+   ? allowPrivateBrowsing: bool .default false,
+}
+
+extensions.ExtensionPath = {
+   type: "path",
+   path: text,
+}
+
+extensions.ExtensionArchivePath = {
+   type: "archive-path",
+   path: text,
+}
+
+extensions.ExtensionBase64Encoded = {
+   type: "base64",
+   value: text,
+}
+</pre>
+
+The <code>extensions.Extension</code> type represents an extension with its additional properties.
+
+<div algorithm>
+To <dfn>expand a install path spec</dfn> given |install path spec|:
+
+1. Let |path| be a value of <code>path</code> field of |install path spec| if the value of <code>type</code> of the |install path spec| is equal to <code>path</code>,
+   otherwise perform any implementation-defined steps to create a temporary folder and make its path to be the value of |path|.
+
+1. Let |archive path| be a value of <code>path</code> field of |install path spec| if the value of <code>type</code> of the |install path spec| is equal to <code>archive-path</code>,
+   otherwise if the value of <code>type</code> of the |install path spec| is equal to <code>archive-path</code> perform any implementation-defined steps to create a temporary file
+   being a result of Base64 decodeing of the value of <code>value</code> filed of |install path spec| and make its path to be the value of |archive path|,
+   otherwise |archive path| is null.
```

Comment:
I do realize this is overly complicated, but wasn't sure how to express it otherwise. Please advise.

Reply:
I'm confused by the flow of this, since each step has something along the lines of "If A do this, otherwise if A do that". I would have expected something like:

1. If the type is "path", return the path.
2. Otherwise, if the type is "archive-path", extract the archive and return the path to the temporary directory.
3. Otherwise, if the type is "base64", decode this as a zip and return the path to the temporary directory.

Reply:
Is it better now? or you would rather remove the intermediate data structure?

Reply:
The extraction process is much clearer now, thanks - I'm happy with that as long as we are comfortable having a switch vs a series of if/else if (cc/ @jgraham). I'm still a little unclear why we need |extension files| and |extension archive|. Is there a reason for that?